src/v/archival/service.cc

```diff
@@ -0,0 +1,378 @@
+              vlog(
+                archival_log.error,
+                "Manifest download for partition {}, failed, "
+                "error: ",
+                ntp.path(),
+                err.what());
+              return ss::now();
+          });
+    });
+}
+
+ss::future<>
+scheduler_service_impl::remove_archivers(std::vector<model::ntp> to_remove) {
+    return ss::parallel_for_each(
+      to_remove, [this](const model::ntp& ntp) -> ss::future<> {
+          vlog(archival_log.info, "removing archiver for {}", ntp.path());
+          auto archiver = _queue[ntp];
+          return ss::with_semaphore(
+                   _conn_limit, 1, [archiver] { return archiver->stop(); })
+            .finally([this, ntp] {
+                vlog(archival_log.info, "archiver stopped {}", ntp.path());
+                _queue.erase(ntp);
+            });
+      });
+}
+
+ss::future<> scheduler_service_impl::reconcile_archivers() {
+    gate_guard g(_gate);
+    cluster::partition_manager& pm = _partition_manager.local();
+    storage::api& api = _storage_api.local();
+    storage::log_manager& lm = api.log_mgr();
+    auto snapshot = lm.get();
+    std::vector<model::ntp> to_remove;
+    std::vector<model::ntp> to_create;
+    // find ntps that exist in _svc_per_ntp but no longer present in log_manager
+    _queue.copy_if(
+      std::back_inserter(to_remove), [&snapshot, &pm](const model::ntp& ntp) {
+          return !snapshot.contains(ntp) || !pm.get(ntp)
+                 || !pm.get(ntp)->is_leader();
+      });
+    // find new ntps that present in the snapshot only
+    std::copy_if(
+      snapshot.begin(),
+      snapshot.end(),
+      std::back_inserter(to_create),
+      [this, &pm](const model::ntp& ntp) {
+          return ntp.ns != model::redpanda_ns && !_queue.contains(ntp)
+                 && pm.get(ntp) && pm.get(ntp)->is_leader();
+      });
+    // epxect to_create & to_remove be empty most of the time
+    if (unlikely(!to_remove.empty() || !to_create.empty())) {
+        // run in parallel
+        ss::gate reconcile_gate;
+        if (!to_remove.empty()) {
+            (void)ss::with_gate(
+              reconcile_gate,
+              [this, to_remove = std::move(to_remove)]() mutable {
+                  return remove_archivers(std::move(to_remove));
+              });
+        }
+        if (!to_create.empty()) {
+            (void)ss::with_gate(
+              reconcile_gate,
+              [this, to_create = std::move(to_create)]() mutable {
+                  return create_archivers(std::move(to_create));
+              });
+        }
+        co_await reconcile_gate.close();
+    }
+    co_return;
+}
+
+ss::future<> scheduler_service_impl::run_uploads() {
+    // Start uploads
+    gate_guard g(_gate);
+    ss::gate upload_gate;
+    ntp_archiver::batch_result total{};
+    int quota = _queue.size();
+    while (_conn_limit.available_units() && quota-- > 0) {
```

Comment:
this is a little hard to read, and i'm wondering if this might be a case where avoiding co-routines makes things simpler. if `archiver->upload_next_candidate(_conn_limit, lm)` grabs a unit of the `_conn_limit` resource semaphore then there is a natural limit imposed rather than having to do accounting here in this loop.

otherwise i think a comment describing how this function works would be helpful as we all start to learn about good coroutine patterns.

from the looks of it the function will return after some work, bound by _conn_limit, has been completed.

now that i'm thinking about it, would it make sense to put this entire loop in a background worker fiber that runs all the time, consuming upload candidates from the queue, and imposing the semaphore conn limit resource restriction?